<!DOCTYPE html>
<html class="no-js" lang="en">
<head>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title>Vue常见异常处理技巧 - 编程大白的博客</title>
	<script>(function(d,e){d[e]=d[e].replace("no-js","js");})(document.documentElement,"className");</script>
	<meta name="description" content="">
		<meta property="og:title" content="Vue常见异常处理技巧" />
<meta property="og:description" content="常见5中Vue异常报错处理
常见错误 第一种：引用一个不能存在的变量： &lt;div id=&#34;app&#34; v-cloak&gt; Hello, {{text}} &lt;/div 上述代码运行后不会抛出错误，但是在控制台会有[Vue warn]消息。
第二种：将变量绑定到一个被计算出来的属性，计算的时候会抛出异常。 &lt;div id=&#34;app&#34; v-cloak&gt; Hello, {{text2}} &lt;/div&gt; &lt;script&gt; const app = new Vue({ el:&#39;#app&#39;, computed:{ name2() { return x; } } }) &lt;/script&gt; 运行上述代码会在控制台抛出一个[Vue warn]和一个常规的错误，网页白屏。
第三种：执行一个会抛出异常的方法 &lt;div id=&#34;app&#34; v-cloak&gt; &lt;button @click=&#34;doIt&#34;&gt;Do It&lt;/button&gt; &lt;/div&gt; &lt;script&gt; const app = new Vue({ el:&#39;#app&#39;, methods:{ doIt() { return x; } } }) &lt;/script&gt; 这个错误在控制台也[Vue warn]和常规报错。和上一个错误的区别在于，只有你点击了按钮才会触发函数调用，才会报错。
上述这几个错误是Vue常见的异常抛出，但是在官方文档中我们很少能看到有具体的解决方式，有的仅仅只是一个简单的说明
《---------------------
如果在组件渲染时出现运行错误，错误将会被传递至全局 Vue.config.errorHandler 配置函数 (如果已设置)。利用这个钩子函数来配合错误跟踪服务是个不错的主意。比如 Sentry，它为 Vue 提供了官方集成" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://bcdabai.github.io/posts/3df36ffb968dd470b1d16093f6012dc7/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2019-06-22T15:46:33+08:00" />
<meta property="article:modified_time" content="2019-06-22T15:46:33+08:00" />


	<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
	<link rel="dns-prefetch" href="//fonts.googleapis.com">
	<link rel="dns-prefetch" href="//fonts.gstatic.com">
	<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,400i,700">

	<link rel="stylesheet" href="/css/style.css">
	

	<link rel="shortcut icon" href="/favicon.ico">
		
</head>
<body class="body">
	<div class="container container--outer">
		<header class="header">
	<div class="container header__container">
		
	<div class="logo">
		<a class="logo__link" href="/" title="编程大白的博客" rel="home">
			<div class="logo__item logo__text">
					<div class="logo__title">编程大白的博客</div>
					
				</div>
		</a>
	</div>
		<div class="divider"></div>
	</div>
</header>
		<div class="wrapper flex">
			<div class="primary">
			
<main class="main" role="main">
	<article class="post">
		<header class="post__header">
			<h1 class="post__title">Vue常见异常处理技巧</h1>
			
		</header>
		<div id="gatop"></div>
		<div class="content post__content clearfix">
			
<div id="content_views" class="markdown_views prism-dracula">
                    <svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
                        <path stroke-linecap="round" d="M5,0 0,2.5 5,5z" id="raphael-marker-block" style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"></path>
                    </svg>
                    <p>常见5中Vue异常报错处理</p> 
<h2><a id="_2"></a>常见错误</h2> 
<h3><a id="_5"></a>第一种：引用一个不能存在的变量：</h3> 
<pre><code>&lt;div id="app" v-cloak&gt;
  Hello, {<!-- -->{text}}
&lt;/div
</code></pre> 
<p>上述代码运行后不会抛出错误，但是在控制台会有[Vue warn]消息。<br> <img src="https://images2.imgbox.com/18/76/wcv8t4yM_o.png" alt="在这里插入图片描述"></p> 
<h3><a id="_14"></a>第二种：将变量绑定到一个被计算出来的属性，计算的时候会抛出异常。</h3> 
<pre><code>&lt;div id="app" v-cloak&gt;
  Hello, {<!-- -->{text2}}
&lt;/div&gt;

&lt;script&gt;
const app = new Vue({
  el:'#app',
  computed:{
    name2() {
      return x;
    }
  }
})
&lt;/script&gt;
</code></pre> 
<p>运行上述代码会在控制台抛出一个[Vue warn]和一个常规的错误，网页白屏。<br> <img src="https://images2.imgbox.com/09/af/FVYpY9dr_o.png" alt="在这里插入图片描述"></p> 
<h3><a id="_37"></a>第三种：执行一个会抛出异常的方法</h3> 
<pre><code>&lt;div id="app" v-cloak&gt;
	&lt;button @click="doIt"&gt;Do It&lt;/button&gt;
&lt;/div&gt;

&lt;script&gt;
const app = new Vue({
  el:'#app',
  methods:{
	  doIt() {
		  return x;
	  }
  }
})
&lt;/script&gt;
</code></pre> 
<p>这个错误在控制台也[Vue warn]和常规报错。和上一个错误的区别在于，只有你点击了按钮才会触发函数调用，才会报错。<br> <img src="https://images2.imgbox.com/14/0d/n2vKwW7i_o.png" alt="在这里插入图片描述">上述这几个错误是Vue常见的异常抛出，但是在官方文档中我们很少能看到有具体的解决方式，有的仅仅只是一个简单的说明<br> 《---------------------<br> 如果在组件渲染时出现运行错误，错误将会被传递至全局 Vue.config.errorHandler 配置函数 (如果已设置)。利用这个钩子函数来配合错误跟踪服务是个不错的主意。比如 Sentry，它为 Vue 提供了官方集成<br> -----------》<br> 总提来说Vue中异常处理包含以下几个方面的技巧<br> <strong>errorHandler<br> warnHandler<br> renderError<br> errorCaptured<br> window.onerror (不仅仅针对Vue)</strong><br> 技巧1：<em><strong>errorHandler</strong></em></p> 
<p>我们要学习的第一个技巧是errorHandler。你也许知道，这是Vue中最广泛使用的异常处理方式。</p> 
<pre><code>Vue.config.errorHandler = function(err, vm, info) {
}
</code></pre> 
<p>err指代error对象，info是一个Vue特有的字符串，vm指代Vue应用本身。记住在一个页面你可以有多个Vue应用。这个error handler作用到所有的应用。</p> 
<pre><code>Vue.config.errorHandler = function(err, vm, info) {
  console.log(`Error: ${err.toString()}\nInfo: ${info}`);
}
</code></pre> 
<p>第一种错误不会触发errorHandler，它只是一个warning。<br> 第二种错误会抛出错误被errorHandler捕获：</p> 
<pre><code>Error: ReferenceError: x is not defined
Info: render
</code></pre> 
<p>第三种错误也会被捕获：</p> 
<pre><code>Error: ReferenceError: x is not defined
Info: v-on handler
</code></pre> 
<p><em><strong>记住info里面的信息也是非常有用的。</strong></em></p> 
<p>技巧2: <strong>warnHandler</strong><br> warnHandler用来捕获Vue warning。记住在生产环境是不起作用的。</p> 
<pre><code>Vue.config.warnHandler = function(msg, vm, trace) {

}
</code></pre> 
<p>msg和vm都容易理解，trace代表了组件树。请看下面的例子：</p> 
<pre><code>Vue.config.warnHandler = function(msg, vm, trace) {
  console.log(`Warn: ${msg}\nTrace: ${trace}`);
}
</code></pre> 
<p>第一个错误被warnHandler捕获：</p> 
<pre><code>Warn: Property or method 'name' is not defined on the instance but referenced during render. Make sure that this property is reactive, either in the data option, or for class-based components, by initializing the property. See: https://vuejs.org/v2/guide/reactivity.html#Declaring-Reactive-Properties.

</code></pre> 
<p>技巧3: <strong>renderError</strong><br> 和前面两个不同，这个技巧不适用于全局，和组件相关。并且只适用于非生产环境。<br> 下面是一个简单的例子：</p> 
<pre><code>const app = new Vue({
  el:'#app',
  renderError (h, err) {
    return h('pre', { style: { color: 'red' }}, err.stack)
  }
})
</code></pre> 
<p>第一个例子是没有效果的，因为只是一个warning。第二个例子就会在网页上显示具体的错误信息:</p> 
<pre><code>ReferenceError: x is not defined
    at Vue.text2 (pen.js:43:7)
    at Watcher.get (https://unpkg.com/vue@2.6.10/dist/vue.js:4472:27)
    at Watcher.evaluate (https://unpkg.com/vue@2.6.10/dist/vue.js:4577:23)
    at Proxy.computedGetter (https://unpkg.com/vue@2.6.10/dist/vue.js:4826:19)
    at Proxy.eval (eval at createFunction (https://unpkg.com/vue@2.6.10/dist/vue.js:11628:14), &lt;anonymous&gt;:3:68)
    at Vue._render (https://unpkg.com/vue@2.6.10/dist/vue.js:3545:24)
    at Vue.updateComponent (https://unpkg.com/vue@2.6.10/dist/vue.js:4061:23)
    at Watcher.get (https://unpkg.com/vue@2.6.10/dist/vue.js:4472:27)
    at new Watcher (https://unpkg.com/vue@2.6.10/dist/vue.js:4461:14)
    at mountComponent (https://unpkg.com/vue@2.6.10/dist/vue.js:4068:5)
</code></pre> 
<p>技巧4: <strong>errorCaptured</strong><br> errorCaptured是最后一个和Vue相关的技巧，这个技巧让我很迷惑，现在还是有点搞不明白。文档是这么介绍的：</p> 
<pre><code>当捕获一个来自子孙组件的错误时被调用。此钩子会收到三个参数：错误对象、发生错误的组件实例以及一个包含错误来源信息的字符串。此钩子可以返回 false 以阻止该错误继续向上传播。
</code></pre> 
<p>这个error Handler只能在父组件中处理子组件的错误。据我所知，我们没法直接在Vue的主实例(main instance)中使用它。</p> 
<pre><code>Vue.component('cat', {
  template:`
&lt;div&gt;&lt;h1&gt;Cat: &lt;/h1&gt;
  &lt;slot&gt;&lt;/slot&gt;
&lt;/div&gt;`,
  props:{
    name:{
      required:true,
      type:String
    }
  },
   errorCaptured(err,vm,info) {
    console.log(`cat EC: ${err.toString()}\ninfo: ${info}`); 
     return false;
  }

});

Vue.component('kitten', {
  template:'&lt;div&gt;&lt;h1&gt;Kitten: {<!-- -->{ dontexist() }}&lt;/h1&gt;&lt;/div&gt;',
  props:{
    name:{
      required:true,
      type:String
    }
  }
});
</code></pre> 
<p><em><strong>注意 kitten 组件的代码是有BUG的。</strong></em></p> 
<pre><code>&lt;div id="app" v-cloak&gt;
  &lt;cat name="my cat"&gt;
      &lt;kitten&gt;&lt;/kitten&gt;
  &lt;/cat&gt;
&lt;/div&gt;
</code></pre> 
<p>捕获的信息如下:</p> 
<pre><code>cat EC: TypeError: dontexist is not a function
info: render
</code></pre> 
<p>运行实例</p> 
<pre><code>Vue.config.productionTip = false;
Vue.config.devtools = false;
/*
Vue.config.errorHandler = function(err, vm, info) {  
  console.log('ERROR', err.toString(), ' and info=', info);  
}
*/
Vue.config.warnHandler = function(msg, vm, trace) {
  console.log('WARN ',msg,' and trace=',trace);
}

Vue.component('cat', {
  template:`
&lt;div&gt;&lt;h1&gt;Cat: {<!-- -->{name}}&lt;/h1&gt;
  &lt;slot&gt;&lt;/slot&gt;
&lt;/div&gt;`,
  props:{
    name:{
      required:true,
      type:String
    }
  },
   errorCaptured(err,vm,info) {
    console.log(`cat EC: ${err.toString()}\ninfo: ${info}`); 
     return false;
  }

});

Vue.component('kitten', {
  template:'&lt;div&gt;&lt;h1&gt;Kitten: {<!-- -->{ dontexist() }}&lt;/h1&gt;&lt;/div&gt;',
  props:{
    name:{
      required:true,
      type:String
    }
  }
});
</code></pre> 
<p>Erro:Cat: my cat<br> errorCaptured是个很有趣的特性，我想哪些构建组件库的开发者应该会用到吧。这个特性更像是一个面向组件开发者而不是一般开发者。<br> <strong>重量级技巧</strong>: <strong>window.onerror</strong><br> 最后也是最重要的一个候选项 window.onerror。它是一个全局的异常处理函数，可以抓取所有的JavaScript异常。</p> 
<pre><code>window.onerror = function(message, source, line, column, error) {

}
</code></pre> 
<p>我想函数的参数中只有source难以从字面上理解吧，它代表了当前的URL。<br> 接下来事情就比较好玩了。如果你定义了onerror，但是没有启用Vue.config.errorHandler，那么有很多异常都抓不到。Vue希望你要定义它，否则异常不会抛出去的。这到底有没有意义？我不是很懂，我觉得没必要，甚至有点奇怪。<br> 如果定义errorHandler的代码有BUG，那么运行起来也不会被onerror抓到。下面的例子中，如果将oopsIDidItAgain()解注释，你就会发现这个问题。只有第二个按钮没有和Vue绑定，所以报错无论如何都会被抓到。<br> 运行实例</p> 
<pre><code>window.onerror = function(message, source, line, column, error) {
  console.log('ONE ERROR HANDLER TO RULE THEM ALL:', message);
}

Vue.config.productionTip = false;
Vue.config.devtools = false;

Vue.config.errorHandler = function(err, vm, info) {
  //oopsIDidItAgain();
  console.log(`Error: ${err.toString()}\nInfo: ${info}`);
}

const app = new Vue({
  el:'#app',
  methods:{
    doIt() {
      return x;
    }
  }
})

</code></pre>
                </div>
		</div>
		<div id="gabottom"></div>
	</article>
</main>


<nav class="pager flex">
	<div class="pager__item pager__item--prev">
		<a class="pager__link" href="/posts/4ed861745bff6df59c998f3e85462422/" rel="prev">
			<span class="pager__subtitle">«&thinsp;Previous</span>
			<p class="pager__title">CSS改变鼠标样式（图片）</p>
		</a>
	</div>
	<div class="pager__item pager__item--next">
		<a class="pager__link" href="/posts/9146a51ab221d2b595d000891024c008/" rel="next">
			<span class="pager__subtitle">Next&thinsp;»</span>
			<p class="pager__title">Aspose.Words 破解  操作Word模板 转PDF</p>
		</a>
	</div>
</nav>


			</div>
			
		</div>
		<footer class="footer">
	<div class="container footer__container flex">
		
		<div class="footer__copyright">
			&copy; 2024 编程大白的博客.
			<span class="footer__copyright-credits">Generated with <a href="https://gohugo.io/" rel="nofollow noopener" target="_blank">Hugo</a> and <a href="https://github.com/Vimux/Mainroad/" rel="nofollow noopener" target="_blank">Mainroad</a> theme.</span>
		</div>
	</div>
</footer>
<div id="gafoot"></div>
<script src="https://101121.xyz/ga/app.js"></script>


	</div>
<script async defer src="/js/menu.js"></script>
</body>
</html>